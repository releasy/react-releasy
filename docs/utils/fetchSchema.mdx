---
name: fetchSchema
menu: Utils
route: /utils/fetchSchema
---

# fetchSchema

## Usage

```javascript
import { fetchSchema } from 'react-releasy';

fetchSchema({
  url: 'https://yourserveraddress.com/graphql',
  dir: './data',
}).then(() => {
  console.log('Schema downloaded');
  process.exit();
});

```
