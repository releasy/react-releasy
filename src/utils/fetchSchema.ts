import {
  buildClientSchema,
  printSchema,
  introspectionQuery,
} from 'graphql';

import { IS_NODE } from '../utils/constants';

type FetchSchemaInput = {
  url: string,
  dir: string,
  filename?: string,
};

type FetchSchemaOutput = {
  path: string,
};

const fetchSchema = async (input: FetchSchemaInput): Promise<FetchSchemaOutput> => {
  if (!IS_NODE) {
    throw new Error('Node.JS environments only!');
  }

  const fs = require('fs');
  const fetchWithRetries = require('fbjs/lib/fetchWithRetries');

  const { url, dir, filename = 'schema.graphql' } = input;

  const response = await fetchWithRetries(url, {
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      query: introspectionQuery,
    }),
  });

  const result = await response.json();

  const clientSchema = buildClientSchema(result.data);
  const schema = printSchema(clientSchema);

  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir);
  }

  const path = `${dir}/${filename}`;

  fs.writeFileSync(path, schema);

  return {
    path,
  };
};

export default fetchSchema;
