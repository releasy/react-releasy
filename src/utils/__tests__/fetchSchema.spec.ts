import fetchSchema from '../fetchSchema';
import fs from 'fs';

it('should fetch the schema', async () => {
  const CONSTANTS = require('../../utils/constants');
  CONSTANTS.IS_NODE = true;

  const json = require('../../../tests/data/schema.json');

  global.fetch.mockResponseOnce(
    JSON.stringify(json),
    {
      status: 200,
      headers: {
        'content-type': 'application/json',
      },
    },
  );

  const result = await fetchSchema({
    url: '/graphql',
    dir: './tests/data',
  });

  const pathExists = fs.existsSync(result.path);

  expect(pathExists).toBe(true);
  expect(result).toMatchSnapshot();
});

it('should create the dir if it doesnt exist', async () => {
  const json = require('../../../tests/data/schema.json');

  global.fetch.mockResponseOnce(
    JSON.stringify(json),
    {
      status: 200,
      headers: {
        'content-type': 'application/json',
      },
    },
  );

  const dir = './tests/non-existing';

  const result = await fetchSchema({
    dir,
    url: '/graphql',
  });

  const dirExists = fs.existsSync(dir);

  expect(dirExists).toBe(true);

  fs.unlinkSync(result.path);
  fs.rmdirSync(dir);
});

it('should thrown an error on a non node environment', () => {
  const CONSTANTS = require('../../utils/constants');
  CONSTANTS.IS_NODE = false;

  const callFetchSchema = () => {
    return fetchSchema({
      url: '/graphql',
      dir: './tests/data',
    });
  };

  expect(callFetchSchema()).rejects.toThrowErrorMatchingSnapshot();
});
