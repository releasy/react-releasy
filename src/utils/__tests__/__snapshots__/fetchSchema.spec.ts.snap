// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fetch the schema 1`] = `
Object {
  "path": "./tests/data/schema.graphql",
}
`;

exports[`should thrown an error on a non node environment 1`] = `"Node.JS environments only!"`;
